File: Backend/src/main/java/com/shubham/FlightManagementSystem/service/AirportService.java
package com.shubham.FlightManagementSystem.service;

import com.shubham.FlightManagementSystem.entity.Airport;
import com.shubham.FlightManagementSystem.repository.AirportRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class AirportService {
    
    @Autowired
    private AirportRepository airportRepository;
    
    public List<Airport> getAllAirports() {
        return airportRepository.findAll();
    }
    
    public Optional<Airport> getAirportById(Long id) {
        return airportRepository.findById(id);
    }
    
    public Optional<Airport> getAirportByCode(String code) {
        return airportRepository.findByCode(code);
    }
    
    public List<Airport> getAirportsByCity(String city) {
        return airportRepository.findByCity(city);
    }
    
    public List<Airport> getAirportsByCountry(String country) {
        return airportRepository.findByCountry(country);
    }
    
    public List<Airport> searchAirports(String keyword) {
        return airportRepository.searchAirports(keyword);
    }
    
    public Airport saveAirport(Airport airport) {
        return airportRepository.save(airport);
    }
    
    public Airport updateAirport(Long id, Airport airportDetails) {
        Optional<Airport> optionalAirport = airportRepository.findById(id);
        if (optionalAirport.isPresent()) {
            Airport airport = optionalAirport.get();
            airport.setCode(airportDetails.getCode());
            airport.setName(airportDetails.getName());
            airport.setCity(airportDetails.getCity());
            airport.setCountry(airportDetails.getCountry());
            return airportRepository.save(airport);
        }
        throw new RuntimeException("Airport not found with id: " + id);
    }
    
    public void deleteAirport(Long id) {
        airportRepository.deleteById(id);
    }
    
    public boolean airportExists(String code) {
        return airportRepository.existsByCode(code);
    }
}

File: Frontend/app/services/airport.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Airport } from '../models/airport.model';

@Injectable({
  providedIn: 'root'
})
export class AirportService {
  private apiUrl = 'http://localhost:8080/api/airports';

  constructor(private http: HttpClient) { }

  getAllAirports(): Observable<Airport[]> {
    return this.http.get<Airport[]>(this.apiUrl);
  }

  getAirportById(id: number): Observable<Airport> {
    return this.http.get<Airport>(`${this.apiUrl}/${id}`);
  }

  getAirportByCode(code: string): Observable<Airport> {
    return this.http.get<Airport>(`${this.apiUrl}/code/${code}`);
  }

  getAirportsByCity(city: string): Observable<Airport[]> {
    return this.http.get<Airport[]>(`${this.apiUrl}/city/${city}`);
  }

  getAirportsByCountry(country: string): Observable<Airport[]> {
    return this.http.get<Airport[]>(`${this.apiUrl}/country/${country}`);
  }

  searchAirports(keyword: string): Observable<Airport[]> {
    return this.http.get<Airport[]>(`${this.apiUrl}/search?keyword=${keyword}`);
  }

  createAirport(airport: Airport): Observable<Airport> {
    return this.http.post<Airport>(this.apiUrl, airport);
  }

  updateAirport(id: number, airport: Airport): Observable<Airport> {
    return this.http.put<Airport>(`${this.apiUrl}/${id}`, airport);
  }

  deleteAirport(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}
